### PR TITLE
Make sending of uncompleted Chain Selection impossible

### LINKED ISSUE #395: Sending Slice with Chain Selection that is not finished doesn't work properly
## Current Behavior

If last Selection added to a Slice is Chain Selection that is not stopped/looped, trying to send Slice causes some errors.

## Expected Behavior

Sending Slice should automatically stop Chain Selection. 

## Steps to Reproduce the Problem

  1. Go to Labelling page.
  2. Add some selections. Make sure that for the last selection you've added you've used Chain Selector.
  3. Do not click Stop/Loop. Click Send Label straight away.
  4. Voi'la, new scan has been loaded, Chain Selector is still active, but clicking on Scan does nothing except for throwing TypeErrors in console:
image

## Additional comment (optional)

Additionally, if Chain Selection is the only Selection present, Send button says 'Nothing to tag' as long as the selection is not stopped. Clicking it causes some errors. IMO it should either say 'Send label' (and upon clicking stop&send that selection) or it should be disabled.

### PR DESCRIPTION
**Pull request by piotrwalkusz1**
Fixes #395 

## Proposed Changes

  - Remove `selectingInProgress` variable from `ChainSelector`
  - Tool is able to prevent scan sending now
  - Remove `deselect()` function

## UI changes (optional)

  - "Send label" button can be disabled 

## Additional comment (optional)

1) The problem occurred because after sending the scan the variable `selectingInProgress` wasn't set to `false`. This shows us that state in Tools is problematic and Tools should be stateless as much as possible. We can manage the state of Tools by functions `onSelect`, `onDeselect`, `onScanSent` but this will be error-prone.
2) While using Chain Tool, "Send label" button toggles enable and disable. It distracts me a bit.

### CODE REVIEW

File: frontend/src/app/pages/marker-page/marker-page.component.html [hunk 1]
```diff
@@ -66,7 +66,7 @@ <h3>Actions</h3>
                 <button mat-raised-button color="primary" (click)="skipScan()">
                     Skip scan
                 </button>
-                <button mat-raised-button color="accent" (click)="sendCompleteLabel()"
+                <button mat-raised-button color="accent" [disabled]="!canSendScan()" (click)="sendCompleteLabel()"
```
Comment (jpowie01) [inline frontend/src/app/pages/marker-page/marker-page.component.html]: What if user won't click "Stop" and move to other Slice or switch Tool? :)

As an end user I may be a little bit confused seeing disabled button without any reason. Maybe a tooltip saying "You've got to end your Chain label first!" would be nice.

**Comment by jpowie01** [comment]
@Darkoon, is it fine for you? You reqested changes but haven’t commented since then. Can we merge it?

**Comment by Darkoon** [comment]
image
It works good but I found some kind of bug. When using chain tool, user can still change current tool resulting in having incomplete chain label (visible on every slice of scan). It will be good to block other tool options when using chain tool, leaving only the option to get back to tool selection.

File: frontend/src/app/components/marker/marker.component.ts [hunk 2]
```diff
@@ -268,14 +268,14 @@ export class MarkerComponent extends ScanViewerComponent implements OnInit {
...
-            if (this.currentTool && !this.currentTool.canUseMouseWheel()) {
+            if (this.currentTool && !this.currentTool.canChangeSlice()) {
                 return;
             }
 
             const sliderValue = wheelEvent.deltaY > 0 ? this.slider.value - 1 : this.slider.value + 1;
 
             if (sliderValue >= this.slider.min && sliderValue <= this.slider.max) {
-                this.tools.forEach((tool) => tool.updateCurrentSlice(sliderValue));
+                this.tools.forEach((selector) => selector.updateCurrentSlice(sliderValue));
```
[the remark this answers is not in the document]
Comment (jpowie01) [reply frontend/src/app/components/marker/marker.component.ts]: @piotrwalkusz1, could you check it?

**Comment by piotrwalkusz1** [comment]
@Darkoon Changing tool while using Chain Tool makes the current chain disappear now. Do you have any other proposals?

**Pull request merged**